Better method call error messages

jackh726 opened the pull request:
Rebase/continuation of #71827

~Based on #92360~
~Based on #93118~

There's a decent description in #71827 that I won't copy here (for now at least)

In addition to rebasing, I've tried to restore most of the original suggestions for invalid arguments. Unfortunately, this does make some of the errors a bit verbose. To fix this will require a bit of refactoring to some of the generalized error suggestion functions, and I just don't have the time to go into it right now.

I think this is in a state that the error messages are overall better than before without a reduction in the suggestions given.

~I've tried to split out some of the easier and self-contained changes into separate commits (mostly in #92360, but also one here). There might be more than can be done here, but again just lacking time.~

r? @estebank as the original reviewer of #71827

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4652640452?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/476c1a3feb70ff01de676cce0a1fb2b5ec11e68c/checks/4652640452/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1027 tests
---
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13442) stdout ----
error[E0308]: arguments to this function are incorrect
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13450:5
   |
9  |     fix(&|y| {
   |
   |
note: cyclic type of infinite size
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13450:9
   |
9  |       fix(&|y| {
   |  _________^
10 | |         // Here, when `x` is called, the parameter `y` is equal to `x`.
11 | |     });
   | |_____^
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           for more information
note: function defined here
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13443:4
   |
   |
2  | fn fix<F>(f: &F)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Some expected error codes were not found: ["E0644"]
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13442)

test result: FAILED. 1002 passed; 1 failed; 24 ignored; 0 measured; 0 filtered out; finished in 9.06s
```

</details>

jackh726 commented:
I will resolve the build failures after first round of review or after #92360 gets merged

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4735206582?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b9e7f9031ec77a52ff693e9bff1bb4a9ad11efe9/checks/4735206582/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1029 tests
---
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477) stdout ----
error[E0308]: arguments to this function are incorrect
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:5
   |
9  |     fix(&|y| {
   |
   |
note: cyclic type of infinite size
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:9
   |
9  |       fix(&|y| {
   |  _________^
10 | |         // Here, when `x` is called, the parameter `y` is equal to `x`.
11 | |     });
   | |_____^
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           for more information
note: function defined here
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13478:4
   |
   |
2  | fn fix<F>(f: &F)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Some expected error codes were not found: ["E0644"]
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477)

test result: FAILED. 1004 passed; 1 failed; 24 ignored; 0 measured; 0 filtered out; finished in 8.52s
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #92883) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            )),
+                        ..
+                    },
+                    _,
+                ) => (*span, Some(of)),
+                hir::ExprKind::Call(hir::Expr { span, .. }, _) => (*span, None),
+                hir::ExprKind::MethodCall(_, span, _, _) => (
+                    *span, None, // methods are never ctors
+                ),
+                k => span_bug!(call_span, "checking argument types on a non-call: `{:?}`", k),
+            };
+            let call_name = match ctor_of {
+                Some(CtorOf::Struct) => "struct",
+                Some(CtorOf::Variant) => "enum variant",
+                None => "function",
+            };
+            if c_variadic && provided_arg_count < minimum_input_count {
+                err_code = "E0060";
+            }
+            let mut err = if minimum_input_count == provided_arg_count {
+                struct_span_err!(
+                    tcx.sess,
+                    error_span,
+                    E0308,
+                    "arguments to this {} are incorrect",
+                    call_name,
+                )
+            } else {
+                tcx.sess.struct_span_err_with_code(
+                    error_span,
+                    &format!(
+                        "this {} takes {}{} but {} {} supplied",
+                        call_name,
+                        if c_variadic { "at least " } else { "" },
+                        potentially_plural_count(minimum_input_count, "argument"),
+                        potentially_plural_count(provided_arg_count, "argument"),
+                        if provided_arg_count == 1 { "was" } else { "were" }
+                    ),
+                    DiagnosticId::Error(err_code.to_owned()),
+                )
```
This could be its own change, as it isn't reliant on the subsequent checks.

jackh726 replied:
In #93118, I make error emission a bit more lazy and more like in this PR. It's not a 100% transplant, but close enough the diff should be nicer.

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    return Some(Issue::Permutation(final_permutation));
+                }
+                return None;
+            };
```
Everything since line 430 could be its own struct and impl, so that we can remove all of this logic out of the diagnostic code and make it clearer what is happening to anyone reading the code as prose.

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            // so that an argument which is already spoken for by the input it's in doesn't
+            // spill over into another similarly typed input
+            // ex:
+            //   fn some_func(_a: i32, _a: i32) {}
```
```suggestion
            //   fn some_func(_a: i32, _b: i32) {}
```

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        if input_ty.is_unit() {
+                            suggested_inputs[input_idx] = Some("()".to_string());
+                        } else {
+                            suggested_inputs[input_idx] = Some(format!("{{{}}}", input_ty));
```
we have some code to suggest default values for *some* types, that could be used here.

jackh726 replied:
Can we leave this for followup, or do you think it needs to be done in this PR?

estebank replied:
Follow up PR is fine.

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            &compatibility_matrix[idx][idx]
+                        {
+                            let provided_ty = final_arg_types[input_idx].map(|ty| ty.0).unwrap();
+                            let cause = &self.misc(provided_args[input_idx].span);
```
We might want to create a specific `ObligationCauseCode` for this case, IIF there's value in that.

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    Some(Issue::Extra(idx)) => {
+                        // Eliminate the argument (without touching any inputs)
+                        let arg_idx = arg_indexes[idx];
+                        let arg_type = if let Some((_, ty)) = final_arg_types[arg_idx] {
+                            format!(" of type {}", ty)
```
Also check that `ty` isn't `_` or `Err`.

estebank replied:
```suggestion
                        let arg_type = if let Some((_, ty)) = final_arg_types[arg_idx] {
                            format!(" of type `{}`", ty)
```

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        ));
+                        suggestion_text = match suggestion_text {
+                            None => Some("remove the extra argument"),
+                            Some(_) => Some("did you mean"),
```
Let's think about an alternative to "did you mean"

jackh726 replied:
Can we leave this as a followup?

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            // As we encounter issues, keep track of what we want to provide for the suggestion
+            let mut suggested_inputs: Vec<Option<String>> = vec![None; minimum_input_count];
+            let mut labels = vec![];
+            let mut suggestion_text = None;
```
Add comment explaining that if multiple changes are replaced with "did you mean"

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4862166294?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/4df1792609a755341836f3d3f1feea916e33b331/checks/4862166294/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1029 tests
---
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477) stdout ----
error[E0308]: arguments to this function are incorrect
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:5
   |
9  |     fix(&|y| {
   |
   |
note: cyclic type of infinite size
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:9
   |
9  |       fix(&|y| {
   |  _________^
10 | |         // Here, when `x` is called, the parameter `y` is equal to `x`.
11 | |     });
   | |_____^
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           for more information
note: function defined here
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13478:4
   |
   |
2  | fn fix<F>(f: &F)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Some expected error codes were not found: ["E0644"]
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477)

test result: FAILED. 997 passed; 1 failed; 31 ignored; 0 measured; 0 filtered out; finished in 8.66s
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4873426741?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/27953a6d3d4a572dfed36d49a1b71066fc9161ec/checks/4873426741/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
F...............................i...............................
failures:

---- [mir-opt] mir-opt/inline/inline-trait-method_2.rs stdout ----
thread '[mir-opt] mir-opt/inline/inline-trait-method_2.rs' panicked at 'Output file `/checkout/src/test/mir-opt/inline/inline_trait_method_2.test2.Inline.after.mir` from test does not exist', src/tools/compiletest/src/runtest.rs:3352:25

---- [mir-opt] mir-opt/issue-62289.rs stdout ----
---- [mir-opt] mir-opt/issue-62289.rs stdout ----
thread '[mir-opt] mir-opt/issue-62289.rs' panicked at 'Output file `/checkout/src/test/mir-opt/issue_62289.test.ElaborateDrops.before.mir` from test does not exist', src/tools/compiletest/src/runtest.rs:3352:25

failures:
    [mir-opt] mir-opt/inline/inline-trait-method_2.rs
    [mir-opt] mir-opt/issue-62289.rs
    [mir-opt] mir-opt/issue-62289.rs

test result: FAILED. 159 passed; 2 failed; 3 ignored; 0 measured; 0 filtered out; finished in 5.36s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/mir-opt" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "mir-opt" "--mode" "mir-opt" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:14:10
```

</details>

estebank commented on src/test/ui/argument-suggestions/basic.stderr:
```diff
@@ -0,0 +1,92 @@
+error[E0308]: arguments to this function are incorrect
+  --> $DIR/basic.rs:20:5
+   |
+LL |     invalid(1.0);
+   |     ^^^^^^^
+   |
+note: expected `u32`, found floating-point number
+  --> $DIR/basic.rs:20:13
+   |
+LL |     invalid(1.0);
+   |             ^^^
+note: function defined here
+  --> $DIR/basic.rs:13:4
+   |
+LL | fn invalid(_i: u32) {}
+   |    ^^^^^^^
```
I think that this case is common enough that we might want to special case it slightly. What do you think of

```
error[E0308]: mismatched types: expected `u32`, found floating-point number
  --> $DIR/basic.rs:20:5
   |
LL |     invalid(1.0);
   |     ------- ^^^ expected `u32`, found floating-point number
   |     |
   |     arguments to this function are incorrect
   |
note: function defined here
  --> $DIR/basic.rs:13:4
   |
LL | fn invalid(_i: u32) {}
   |    ^^^^^^^
```

jackh726 replied:
The output has changed a bit (the "expected" note is now a label again). I'm on the fence here: I do think we can have a little bit more specific error message when there is only one problem. However, getting this to be "just right" is probably pretty subjective and I don't really want to put in a ton of time on this or have this sit.

estebank commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -16,18 +18,41 @@ use rustc_hir as hir;
 use std::iter;
 use std::slice;
 
+#[derive(Clone, Debug)]
+pub(crate) enum Compatibility<'tcx> {
+    Compatible,
+    // Coerce error and subtyping error
+    Incompatible(Option<TypeError<'tcx>>, Option<TypeError<'tcx>>),
+}
+
+/// Similar to `Issue`, but contains some extra information
+pub(crate) enum Error<'tcx> {
+    /// The given argument is the invalid type for the input
+    Invalid(usize, Compatibility<'tcx>),
+    /// There is a missing input
+    Missing(usize),
+    /// There's a superfluous argument
+    Extra(usize),
+    /// Two arguments should be swapped
+    Swap(usize, usize, usize, usize),
+    /// Several arguments should be reordered
+    Permutation(Vec<(usize, usize)>), // dest_arg, dest_input
+}
```
Maybe we should move these to `arg_matrix.rs` too.

estebank commented on src/test/ui/argument-suggestions/basic.stderr:
```diff
@@ -0,0 +1,92 @@
+  --> $DIR/basic.rs:21:5
+   |
+LL |     extra("");
+   |     ^^^^^ -- argument unexpected
```
This should be

```suggestion
   |     ----- ^^ unexpected argument
```

jackh726 replied:
Is there any way to do this other than changing the primary span of the error and the span of the label? (Though, then the message would be different...)

estebank commented on src/test/ui/argument-suggestions/basic.stderr:
```diff
@@ -0,0 +1,92 @@
+   |
+LL | fn extra() {}
+   |    ^^^^^
+help: remove the extra argument
+   |
+LL |     extra();
+   |     ~~~~~~~
```
Ideally, given the conversation we had yesterday, I would like this to be, but it doesn't have to be in this PR

```suggestion
help: remove the extra argument
   |
LL -     extra("");
LL +     extra();
   |
```

jackh726 replied:
Let's leave this for followup.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4873986376?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/1a4b47c1e451bd84f9ddafa223dab210acb45bff/checks/4873986376/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1029 tests
---
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477) stdout ----
error[E0308]: arguments to this function are incorrect
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:5
   |
9  |     fix(&|y| {
   |
   |
note: cyclic type of infinite size
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:9
   |
9  |       fix(&|y| {
   |  _________^
10 | |         // Here, when `x` is called, the parameter `y` is equal to `x`.
11 | |     });
   | |_____^
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           for more information
note: function defined here
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13478:4
   |
   |
2  | fn fix<F>(f: &F)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Some expected error codes were not found: ["E0644"]
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477)

test result: FAILED. 997 passed; 1 failed; 31 ignored; 0 measured; 0 filtered out; finished in 8.85s
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4875254854?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e789f0fdcde1e2d8f50b8e5699be09bcbfb86be7/checks/4875254854/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1029 tests
---
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477) stdout ----
error[E0308]: arguments to this function are incorrect
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:5
   |
9  |       fix(&|y| {
   |  _____^^^_-
10 | |         // Here, when `x` is called, the parameter `y` is equal to `x`.
11 | |     });
   | |_____- cyclic type of infinite size
   |
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           for more information
note: function defined here
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13478:4
   |
   |
2  | fn fix<F>(f: &F)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Some expected error codes were not found: ["E0644"]
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477)

test result: FAILED. 997 passed; 1 failed; 31 ignored; 0 measured; 0 filtered out; finished in 7.94s
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4887537386?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/113b0657a976c174627a8262f2def621b1e89db6/checks/4887537386/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1029 tests
---
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477) stdout ----
error[E0308]: arguments to this function are incorrect
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:5
   |
9  |       fix(&|y| {
   |  _____^^^_-
10 | |         // Here, when `x` is called, the parameter `y` is equal to `x`.
11 | |     });
   | |_____- cyclic type of infinite size
   |
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           for more information
note: function defined here
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13478:4
   |
   |
2  | fn fix<F>(f: &F)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Some expected error codes were not found: ["E0644"]
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477)

test result: FAILED. 997 passed; 1 failed; 31 ignored; 0 measured; 0 filtered out; finished in 8.86s
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4888771783?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/cfed042441eb8fe023e0141f47274f218b403547/checks/4888771783/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1029 tests
---
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477) stdout ----
error[E0308]: arguments to this function are incorrect
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13485:5
   |
9  |       fix(&|y| {
   |  _____^^^_-
10 | |         // Here, when `x` is called, the parameter `y` is equal to `x`.
11 | |     });
   | |_____- cyclic type of infinite size
   |
   = note: closures cannot capture themselves or take themselves as argument;
           this error may be the result of a recent compiler bug-fix,
           for more information
note: function defined here
  --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:13478:4
   |
   |
2  | fn fix<F>(f: &F)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Some expected error codes were not found: ["E0644"]
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0644 (line 13477)

test result: FAILED. 997 passed; 1 failed; 31 ignored; 0 measured; 0 filtered out; finished in 9.05s
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4890028891?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/d9725e93d1ed0d3f3a2cba2edf3409070c6d9b52/checks/4890028891/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

running 229 tests
..........i.ii....ii................................................................................ 100/229
.................i...................iiiiiii......i...................iii........................... 200/229
Some tests failed in compiletest suite=run-make-fulldeps mode=run-make host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
...........ii.F..............

---- [run-make] run-make-fulldeps/type-mismatch-same-crate-name stdout ----

error: make failed
error: make failed
status: exit status: 2
command: "make"
stdout:
------------------------------------------
# compile two different versions of crateA
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  --crate-type=rlib crateA.rs -C metadata=-1 -C extra-filename=-1
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  --crate-type=rlib crateA.rs -C metadata=-2 -C extra-filename=-2
# make crateB depend on version 1 of crateA
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  --crate-type=rlib crateB.rs --extern crateA=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name/libcrateA-1.rlib
# make crateC depend on version 2 of crateA
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  crateC.rs --extern crateA=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name/libcrateA-2.rlib 2>&1 | \
 tr -d '\r\n' | "/checkout/src/etc/cat-and-grep.sh" -e \
"mismatched types.*\
crateB::try_foo\(foo2\);.*\
expected struct \`crateA::foo::Foo\`, found struct \`Foo\`.*\
different versions of crate \`crateA\`.*\
mismatched types.*\
crateB::try_bar\(bar2\);.*\
expected trait \`crateA::bar::Bar\`, found trait \`Bar\`.*\
different versions of crate \`crateA\`"
[[[ begin stdout ]]]
error[E0308]: arguments to this function are incorrect  --> crateC.rs:22:9   |22 |         crateB::try_foo(foo2);   |         ^^^^^^^^^^^^^^^ ---- expected struct `crateA::foo::Foo`, found struct `Foo`   |   = note: perhaps two different versions of crate `crateA` are being used?error[E0308]: arguments to this function are incorrect  --> crateC.rs:23:9   |23 |         crateB::try_bar(bar2);   |         ^^^^^^^^^^^^^^^ ---- expected trait `crateA::bar::Bar`, found trait `Bar`   |   = note: expected struct `Box<(dyn crateA::bar::Bar + 'static)>`              found struct `Box<dyn Bar>`   = note: perhaps two different versions of crate `crateA` are being used?error: aborting due to 2 previous errorsFor more information about this error, try `rustc --explain E0308`.
[[[ end stdout ]]]
Error: cannot match: mismatched types.*crateB::try_foo\(foo2\);.*expected struct `crateA::foo::Foo`, found struct `Foo`.*different versions of crate `crateA`.*mismatched types.*crateB::try_bar\(bar2\);.*expected trait `crateA::bar::Bar`, found trait `Bar`.*different versions of crate `crateA`
------------------------------------------
stderr:
------------------------------------------
warning: trait objects without an explicit `dyn` are deprecated
warning: trait objects without an explicit `dyn` are deprecated
 --> crateA.rs:8:25
8 |     pub fn bar() -> Box<Bar> {
  |                         ^^^
  |
  = note: `#[warn(bare_trait_objects)]` on by default
---

warning: 1 warning emitted

warning: trait objects without an explicit `dyn` are deprecated
 --> crateA.rs:8:25
8 |     pub fn bar() -> Box<Bar> {
  |                         ^^^
  |
  = note: `#[warn(bare_trait_objects)]` on by default
---

warning: trait objects without an explicit `dyn` are deprecated
 --> crateB.rs:4:23
  |
4 | pub fn try_bar(x: Box<crateA::Bar>){}
  |
  = note: `#[warn(bare_trait_objects)]` on by default
  = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
  = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
  = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
help: use `dyn`
  |
4 - pub fn try_bar(x: Box<crateA::Bar>){}
4 + pub fn try_bar(x: Box<dyn crateA::Bar>){}

warning: unused variable: `x`
 --> crateB.rs:3:16
  |
  |
3 | pub fn try_foo(x: crateA::Foo){}
  |                ^ help: if this is intentional, prefix it with an underscore: `_x`
  = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `x`
 --> crateB.rs:4:16
 --> crateB.rs:4:16
  |
4 | pub fn try_bar(x: Box<crateA::Bar>){}
  |                ^ help: if this is intentional, prefix it with an underscore: `_x`
warning: 3 warnings emitted


make: *** [Makefile:10: all] Error 1
------------------------------------------




failures:
    [run-make] run-make-fulldeps/type-mismatch-same-crate-name

test result: FAILED. 209 passed; 1 failed; 19 ignored; 0 measured; 0 filtered out; finished in 35.74s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--rust-demangler-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" "--src-base" "/checkout/src/test/run-make-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "run-make-fulldeps" "--mode" "run-make" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--llvm-bin-dir" "/usr/lib/llvm-12/bin" "--cc" "cc" "--cxx" "c++" "--cflags" "-ffunction-sections -fdata-sections -fPIC -m64" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:35:17
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #92787) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4925111852?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/d7819db21cf37294462b985e7f4dbec320c1bf6b/checks/4925111852/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
status: exit status: 2
command: "make"
stdout:
------------------------------------------
# compile two different versions of crateA
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  --crate-type=rlib crateA.rs -C metadata=-1 -C extra-filename=-1
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  --crate-type=rlib crateA.rs -C metadata=-2 -C extra-filename=-2
# make crateB depend on version 1 of crateA
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  --crate-type=rlib crateB.rs --extern crateA=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name/libcrateA-1.rlib
# make crateC depend on version 2 of crateA
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name  crateC.rs --extern crateA=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/type-mismatch-same-crate-name/type-mismatch-same-crate-name/libcrateA-2.rlib 2>&1 | \
 tr -d '\r\n' | "/checkout/src/etc/cat-and-grep.sh" -e \
"arguments to this function are incorrect.*\
crateB::try_foo\(foo2\);.*\
expected struct \`crateA::foo::Foo\`, found struct \`Foo\`.*\
different versions of crate \`crateA\`.*\
arguments to this function are incorrect.*\
crateB::try_bar\(bar2\);.*\
expected trait \`crateA::bar::Bar\`, found trait \`Bar\`.*\
different versions of crate \`crateA\`"
[[[ begin stdout ]]]
error[E0308]: mismatched types  --> crateC.rs:22:25   |22 |         crateB::try_foo(foo2);   |         --------------- ^^^^ expected struct `crateA::foo::Foo`, found struct `Foo`   |         |   |         arguments to this function are incorrect   |   = note: perhaps two different versions of crate `crateA` are being used?error[E0308]: mismatched types  --> crateC.rs:23:25   |23 |         crateB::try_bar(bar2);   |         --------------- ^^^^ expected trait `crateA::bar::Bar`, found trait `Bar`   |         |   |         arguments to this function are incorrect   |   = note: expected struct `Box<(dyn crateA::bar::Bar + 'static)>`              found struct `Box<dyn Bar>`   = note: perhaps two different versions of crate `crateA` are being used?error: aborting due to 2 previous errorsFor more information about this error, try `rustc --explain E0308`.
[[[ end stdout ]]]
Error: cannot match: arguments to this function are incorrect.*crateB::try_foo\(foo2\);.*expected struct `crateA::foo::Foo`, found struct `Foo`.*different versions of crate `crateA`.*arguments to this function are incorrect.*crateB::try_bar\(bar2\);.*expected trait `crateA::bar::Bar`, found trait `Bar`.*different versions of crate `crateA`
------------------------------------------
stderr:
------------------------------------------
warning: trait objects without an explicit `dyn` are deprecated
warning: trait objects without an explicit `dyn` are deprecated
 --> crateA.rs:8:25
8 |     pub fn bar() -> Box<Bar> {
  |                         ^^^
  |
  = note: `#[warn(bare_trait_objects)]` on by default
---

warning: 1 warning emitted

warning: trait objects without an explicit `dyn` are deprecated
 --> crateA.rs:8:25
8 |     pub fn bar() -> Box<Bar> {
  |                         ^^^
  |
  = note: `#[warn(bare_trait_objects)]` on by default
---

warning: trait objects without an explicit `dyn` are deprecated
 --> crateB.rs:4:23
  |
4 | pub fn try_bar(x: Box<crateA::Bar>){}
  |
  = note: `#[warn(bare_trait_objects)]` on by default
  = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
  = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
  = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
help: use `dyn`
  |
4 - pub fn try_bar(x: Box<crateA::Bar>){}
4 + pub fn try_bar(x: Box<dyn crateA::Bar>){}

warning: unused variable: `x`
 --> crateB.rs:3:16
  |
  |
3 | pub fn try_foo(x: crateA::Foo){}
  |                ^ help: if this is intentional, prefix it with an underscore: `_x`
  = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `x`
 --> crateB.rs:4:16
 --> crateB.rs:4:16
  |
4 | pub fn try_bar(x: Box<crateA::Bar>){}
  |                ^ help: if this is intentional, prefix it with an underscore: `_x`
warning: 3 warnings emitted


make: *** [Makefile:10: all] Error 1
------------------------------------------




failures:
    [run-make] run-make-fulldeps/type-mismatch-same-crate-name

test result: FAILED. 209 passed; 1 failed; 19 ignored; 0 measured; 0 filtered out; finished in 31.80s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--rust-demangler-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" "--src-base" "/checkout/src/test/run-make-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "run-make-fulldeps" "--mode" "run-make" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--llvm-bin-dir" "/usr/lib/llvm-12/bin" "--cc" "cc" "--cxx" "c++" "--cflags" "-ffunction-sections -fdata-sections -fPIC -m64" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:34:44
```

</details>

estebank commented on src/test/ui/typeck/struct-enum-wrong-args.stderr:
```diff
@@ -2,95 +2,125 @@ error[E0061]: this enum variant takes 1 argument but 2 arguments were supplied
 error[E0061]: this enum variant takes 1 argument but 3 arguments were supplied
   --> $DIR/struct-enum-wrong-args.rs:7:13
    |
 LL |     let _ = Ok(3, 6, 2);
-   |             ^^ -  -  - supplied 3 arguments
-   |             |
-   |             expected 1 argument
+   |             ^^    -  - argument unexpected
+   |                   |
+   |                   argument unexpected
+   |
+help: did you mean
+   |
+LL |     let _ = Ok(3);
+   |             ~~~~~
```
These changes look great!

I think we could potentially change the wording in the future so we don't talk about arguments and instead talk about fields, but that's independent of this PR.

estebank commented on src/test/ui/typeck/issue-89856.stderr:
```diff
@@ -2,10 +2,17 @@ error[E0308]: mismatched types
    |
    = note: expected enum `Option<&str>`
               found enum `Option<&String>`
+note: function defined here
+  --> $DIR/issue-89856.rs:1:4
+   |
+LL | fn take_str_maybe(x: Option<&str>) -> Option<&str> { None }
+   |    ^^^^^^^^^^^^^^
```
This makes me think: could we special case this to point at the `fn` args here? Potential future change.

estebank commented on src/test/ui/type/type-ascription-instead-of-initializer.stderr:
```diff
@@ -11,15 +11,12 @@ error[E0061]: this function takes 1 argument but 2 arguments were supplied
-note: associated function defined here
-  --> $SRC_DIR/alloc/src/vec/mod.rs:LL:COL
+help: remove the extra argument
    |
-LL |     pub fn with_capacity(capacity: usize) -> Self {
-   |            ^^^^^^^^^^^^^
+LL |     let x: Vec::with_capacity(10);
+   |            ~~~~~~~~~~~~~~~~~~~~~~
```
We are losing the span pointing at the associated fn here. Note to self: check where that regression might be happening.

jackh726 replied:
That would be because somehow we changed from `def_ident_span` to `get_if_local`. Fixing...

estebank commented on src/test/ui/span/missing-unit-argument.stderr:
```diff
@@ -2,88 +2,101 @@ error[E0061]: this enum variant takes 1 argument but 0 arguments were supplied
 
 error[E0061]: this function takes 2 arguments but 0 arguments were supplied
   --> $DIR/missing-unit-argument.rs:12:5
    |
 LL |     foo();
-   |     ^^^-- supplied 0 arguments
+   |     ^^^--
    |     |
-   |     expected 2 arguments
+   |     an argument of type `()` is missing
+   |     an argument of type `()` is missing
    |
```
Should we special case this to provide a single `span_label` saying "two arguments of type `()` and `()` are missing"?

jackh726 replied:
Good idea. Might be worth thinking about consecutive missing args in general:
```
an argument of type `X` is missing
``` 
```
two arguments of type `X` and `Y` are missing
```
```
three arguments of type `X`, `Y`, and `Z` are missing
```
```
multiple missing arguments
```
(the last one is still useful because of the "function defined here" note

I think we only need to fall back here for *consecutive* arguments; if missing args are interspersed by others, the spans will be different and it won't be confusing

We also can probably have a better span here: ideally probably an "empty" span between the parantheses.

estebank replied:
What I like about the current output is that it is simple and effective regardless of count: "expected 2 arguments"/"supplied 0 arguments" doesn't need that much logic to work well in all cases.

An empty span between the parentheses will show up in a problematic way, because it would display as an underline under the closing `)`, and we wouldn't want that. But it can point at the `()`, like it does now.

jackh726 replied:
I'll play around a bit with this. I though you could point at an "empty" span, with just like a line.

jackh726 commented:
Okay @estebank, I think I'm caught up with reviews that need to be done in this PR. This is ready for another round of review. (I'm going to squash the commits before merge, but leaving them for now for better review)

A brief list of reviews made but not changed in this PR:
- Suggest default values instead of e.g. `{i32}`
- Create an ObligationCauseCode instead of misc (don't think this is needed right now)
- Change "did you mean" to something else
  - I did make this get emitted less; now if there are multiple errors of the same type, then that just get pluralized
- "diff" view of removal/addition in error
- Talk about fields instead of arguments for tuple structs/enums
- Pointing at fn args in "function defined here"

A brief list of "big" changes made since reviews:
- ArgMatrix is now in a separate file with it's logic basically completely separate
- We don't emit "of type" if type is error or has inference vars
- If there is a single error, we emit that as we did before
- We merge adjacent missing argument errors into a single one

bors commented:
:umbrella: The latest upstream changes (presumably #90677) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Quantumplation commented:
Hey @jackh726, thanks for taking this and helping update it, really really appreciate that! I've been busy launching a startup, and that's been *super* crazy, so I hadn't had time to revisit.

I'd still like to help out, if there are some small things that I could help with!

jackh726 commented:
At this point, I think this waiting for another round of review from @estebank.

There are some followup items above that can be done "in parallel" as a separate PR based on this - and just rebased when this lands.

estebank commented on src/test/ui/span/coerce-suggestions.stderr:
```diff
@@ -20,19 +20,33 @@ error[E0308]: mismatched types
   --> $DIR/coerce-suggestions.rs:12:10
    |
 LL |     test(&y);
-   |          ^^ types differ in mutability
+   |     ---- ^^ types differ in mutability
+   |     |
+   |     arguments to this function are incorrect
```
Note to self: we should suggest `&mut y` here.

estebank commented on src/test/ui/span/E0057.stderr:
```diff
@@ -2,17 +2,25 @@ error[E0057]: this function takes 1 argument but 0 arguments were supplied
+help: provide the argument
+   |
+LL |     let a = f({_});
+   |             ~~~~~~
```
Somewhat surprised we don't point at the definition of `f` here.

estebank commented on src/test/ui/resolve/resolve-primitive-fallback.stderr:
```diff
@@ -19,15 +19,17 @@ error[E0061]: this function takes 0 arguments but 1 argument was supplied
   --> $DIR/resolve-primitive-fallback.rs:3:5
    |
 LL |     std::mem::size_of(u16);
-   |     ^^^^^^^^^^^^^^^^^ --- supplied 1 argument
-   |     |
-   |     expected 0 arguments
+   |     ^^^^^^^^^^^^^^^^^ --- argument unexpected
    |
 note: function defined here
   --> $SRC_DIR/core/src/mem/mod.rs:LL:COL
    |
 LL | pub const fn size_of<T>() -> usize {
    |              ^^^^^^^
+help: remove the extra argument
+   |
+LL |     std::mem::size_of();
+   |     ~~~~~~~~~~~~~~~~~~~
```
Future work: special case support for noticing that a function argument was supposed to be a type param.

estebank commented on src/test/ui/methods/method-call-err-msg.stderr:
```diff
@@ -2,43 +2,49 @@ error[E0061]: this function takes 0 arguments but 1 argument was supplied
 LL |     fn two(self, _: isize, _: isize) -> Foo { self }
-   |        ^^^ ----  --------  --------
+   |        ^^^
+help: provide the argument
+   |
+LL |      .two(0, {isize})(0);
+   |       ~~~~~~~~~~~~~~~
```
**Please fix** there's a bug in the span here, it should be replacing the `(0)` as well, and it isn't.

estebank commented on src/test/ui/methods/method-call-err-msg.stderr:
```diff
@@ -2,43 +2,49 @@ error[E0061]: this function takes 0 arguments but 1 argument was supplied
 LL |     fn one(self, _: isize) -> Foo { self }
-   |        ^^^ ----  --------
+   |        ^^^
+help: provide the argument
+   |
+LL |      .one({isize})()
+   |       ~~~~~~~~~~~~
```
Same here. I'm guessing there's a small issue with the span used for method calls.

estebank commented on src/test/ui/indexing-requires-a-uint.stderr:
```diff
@@ -11,12 +11,15 @@ error[E0308]: mismatched types
-help: you can convert a `usize` to an `isize` and panic if the converted value doesn't fit
+note: function defined here
+  --> $DIR/indexing-requires-a-uint.rs:5:8
    |
-LL |     bar::<isize>(i.try_into().unwrap());  // i should not be re-coerced back to an isize
-   |                   ++++++++++++++++++++
+LL |     fn bar<T>(_: T) {}
+   |        ^^^
 
```
We lost this suggestion here? Could you check if we can get it back? Not a dealbreaker.

estebank commented on src/test/ui/hrtb/issue-58451.stderr:
```diff
@@ -2,15 +2,19 @@ error[E0061]: this function takes 1 argument but 0 arguments were supplied
 LL | fn f<I>(i: I)
-   |    ^    ----
+   |    ^
+help: provide the argument
+   |
+LL |     f(&[f({_})]);
+   |         ~~~~~~
```
I'm worried that the `{_}` placeholder might be a bit abstract, but we can iterate on this after landing the PR.

estebank commented on src/test/ui/fn/fn-item-type.rs:
```diff
@@ -45,7 +45,7 @@ fn main() {
     eq(foo::<u8>, bar::<u8> as fn(isize) -> isize);
     //~^ ERROR mismatched types
-    //~| expected fn item `fn(_) -> _ {foo::<u8>}`
+    //~| expected type `fn(_) -> _ {foo::<u8>}`
```
Slightly surprised at this change.

estebank reviewed (approved):
I'm ok with the current changes, modulo the invalid suggestions. There's some outstanding work to bring this to par, but lets land this asap.

r=me after fixing the span for incorrect method args suggestion.

jackh726 commented:
I've rebased and fixed the suggestions. Rebasing over the tuple suggestion changes in #91530 was a bit difficult, and the code is a bit duplicated. But I'd rather land this now and cleanup in-tree, rather than risk bitrotting even more.

@bors r=estebank p=1 rollup=never

bors commented:
:pushpin: Commit b6c87c555f3b664f558d137d5f7696a07488157e has been approved by `estebank`

<!-- @bors r=estebank b6c87c555f3b664f558d137d5f7696a07488157e -->
<!-- homu: {"type":"Approved","sha":"b6c87c555f3b664f558d137d5f7696a07488157e","approver":"estebank"} -->

jackh726 commented:
@bors p=10 (bitrotty and I want this to land before the rollup)

bors commented:
:hourglass: Testing commit b6c87c555f3b664f558d137d5f7696a07488157e with merge 07bb916d44a66d2caba427c7ee132bbeb245977b...
<!-- homu: {"type":"BuildStarted","head_sha":"b6c87c555f3b664f558d137d5f7696a07488157e","merge_sha":"07bb916d44a66d2caba427c7ee132bbeb245977b"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6046749189?check_suite_focus=true)
Approved by: estebank
Pushing 07bb916d44a66d2caba427c7ee132bbeb245977b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"estebank","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6046749189?check_suite_focus=true"},"merge_sha":"07bb916d44a66d2caba427c7ee132bbeb245977b"} -->

rust-timer commented:
Finished benchmarking commit (07bb916d44a66d2caba427c7ee132bbeb245977b): [comparison url](https://perf.rust-lang.org/compare.html?start=080d5452e1bb6e18e12a073d4d0283fd9b6dac0b&end=07bb916d44a66d2caba427c7ee132bbeb245977b).

**Summary**: 
- Primary benchmarks: 🎉 relevant improvements found
- Secondary benchmarks: 😿 relevant regressions found

| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 0 | 4 | 3 | 0 | 3 |
| mean[^2] | N/A | 0.8% | -1.1% | N/A | -1.1% |
| max | N/A | 1.2% | -1.5% | N/A | -1.5% |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

pnkfelix commented:
Visiting for weekly performance triage.

* The main regression here was to externs but there's a lot of historical noise in the data for externs, and not trustworthy.

@rustbot label: -perf-regression-triaged

flip1995 commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -327,15 +376,82 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            errors.drain_filter(|error| {
+                let Error::Invalid(input_idx, Compatibility::Incompatible(error)) = error else { return false };
+                let expected_ty = expected_input_tys[*input_idx];
+                let provided_ty = final_arg_types[*input_idx].map(|ty| ty.0).unwrap();
```
@jackh726 This line causes an ICE when compiling `chalk` with `cargo check --all-targets --all-features`. AFAICT the issue here is that `input_idx` is in the range of `0..expected_input_tys.len()`, but the amount of passed types could be smaller. I don't have a reproducer yet. 

I don't think returning the `input_idx` based on the `final_arg_types.len()` would fix the issue, but rather reverse the issue with `expected_input_tys`, if more arguments are passed than expected.

Maybe this `Error` kind should hold both indices :thinking:

flip1995 replied:
Minimal reproducer:

```rust
fn f(_: usize, _: &usize, _: usize) {}

fn arg<T>() -> T { todo!() }

fn main() {
    let x = arg(); // `x` must be inferred
    f(&x, ""); // The reference on `&x` is important to reproduce the ICE
}
```

flip1995 replied:
Opened https://github.com/rust-lang/rust/issues/96638
